dvc/scm/git/__init__.py

```diff
@@ -71,16 +72,31 @@ def clone(url, to_path, rev=None):
             # [1] https://github.com/gitpython-developers/GitPython/issues/924
             env[ld_key] = ""
 
+        clone_from = partial(
+            git.Repo.clone_from,
+            url,
+            to_path,
+            env=env,  # needed before we can fix it in __init__
+            no_single_branch=True,
+        )
+
         try:
-            tmp_repo = git.Repo.clone_from(
-                url,
-                to_path,
-                env=env,  # needed before we can fix it in __init__
-                no_single_branch=True,
-            )
-            tmp_repo.close()
+            if rev:
+                logger.debug("attempting shallow clone of branch %s", rev)
+                tmp_repo = clone_from(branch=rev, depth=1)
+            else:
+                logger.debug("full clone")
+                tmp_repo = clone_from()
```

Comment:
This is very dangerous code, which will cause us grief. We shouldn't hide that we are sometimes make shallow code and sometimes not from outside. Outside may rely on clone being full as it currently does in `external_repo.py`.

Reply:
we're going to have to fix "outside" (i.e. `external_repo.py`) first